client/service.go

```diff
@@ -71,6 +71,8 @@ type Service struct {
 
 	backoff *backoff.Backoff
 
+	subscriptions []packet.Subscription
```

Comment:
In order to keep the order of subscriptions, and let mqtt server handle the overlap or identical topics. If you think the order is not important, I can switch back to map.

Reply:
I guess the order is secondary. And if it's important it can be easily implemented on top. Instead of using a `map` I would actually tend to use a `topic.Tree`. This would efficiently scale if a user has lots of subscriptions.

Reply:
I have read the topic.Tree and it can be used. And the client will handle the identical topics case, for example, user subscribes topic A with QoS 0, then subscribes topic A with QoS 1, the first subscription will be removed.